readcomb/phase_change_filter.py

```diff
@@ -42,151 +34,270 @@ def check_snps(f_name, chromosome, left_bound, right_bound):
+    record: bam record from pysam alignment file
+    ref: reference sequence that is aligned with the query sequence obtained from a fasta file
+    
+    Returns:
+    ref: modified reference sequence that was given with gaps if there are insertions in the query sequence
+    segment: dna sequence that is built using the bam cigar tuples and query segment
+    
+    Function for human readable version of mate pair sequence analysis. Takes in a bam record and a reference
+    sequence and builds the query segment using the index and cigar tuple given by the bam record
+    '''
+    cigar_tuples = record.cigartuples
     
-    # otherwise -
+    # initialize segment for building
+    segment = ''
+    
+    # index to keep track of where we are in the query_segment
+    query_segment = record.query_sequence
     index = 0
     
+    # no alignment, cigar_tuple is equal to None
+    if cigar_tuples == None:
+        return segment
+
     for cigar_tuple in cigar_tuples:
         # 4 = soft clipping, the record.query_sequence has the portion that is soft clipping
         # so we need to skip it with index
         if cigar_tuple[0] == 4:
             index += cigar_tuple[1]
-            
+
         # 5 = hard clipping, record.query_sequence does not have the portion that is
         # hard clipping so we don't skip it and we don't add anything
         elif cigar_tuple[0] == 5:
             continue
-        
-        # if it is a match (0) or an insertion, then just add it onto the segment 
-        elif cigar_tuple[0] in [0, 1]:
-            segment += query_sequence[index:index+cigar_tuple[1]]
+
+        # 0 is a match, just add it onto the segment 
+        elif cigar_tuple[0] == 0:
+            segment += query_segment[index:index+cigar_tuple[1]]
             index += cigar_tuple[1]
-        
+
+        # 1 is an insertion, treated same as a match
+        elif cigar_tuple[0] == 1:
+            segment += query_segment[index:index+cigar_tuple[1]]
+            index += cigar_tuple[1]\
+
+        # 2 is an insertion, add a gap to the query
+        elif cigar_tuple[0] == 2:
+            segment += '-' * cigar_tuple[1]
+            index += cigar_tuple[1]
+
         else:
-            print('oops forgot to consider this: ' + str(cigar_tuple))
+            print('forgot to consider this: ' + str(cigar_tuple))
             print(cigar_tuples)
-
+    
     return segment
 
 
-def parse_phase_changes(snps, cigar_tuples, segment, record):
-
-    no_match_flag = False
-    phase_change_flag = False
-    previous_strand = None
+def phase_detection(snps, segment, record):
+    
+    '''
+    snps: list of snps in the area of the sequence and record given by the function check_snps(),
+        each snp is a cyvcf2 variant object
+    segment: string sequence that is the segment built by the function 
+    '''
+    
     snp_lst = []
     
+    cigar_tuples = record.cigartuples
+    
+    # check for no alignment
+    if cigar_tuples == None:
```

Comment:
TODO: find read raising this issue

Reply:
Created text file called unaligned_reads.txt with around 360 unaligned reads that you could look at, will be included in the next commit